crates/core_simd/src/array.rs

```diff
@@ -0,0 +1,165 @@
+    SimdUsize<LANES>: crate::LanesAtMost32,
+    SimdIsize<LANES>: crate::LanesAtMost32,
+    MaskSize<LANES>: crate::Mask,
+    Self: Sized,
+{
+    /// The scalar type in every lane of this vector type.
+    type Scalar: Copy + Sized;
+
+    /// Generates a SIMD vector with the same value in every lane.
+    #[must_use]
+    fn splat(val: Self::Scalar) -> Self;
+
+    /// SIMD gather: construct a SIMD vector by reading from a slice, using potentially discontiguous indices.
+    /// Out-of-bounds indices instead use the default value (zero, generally).
+    /// ```
+    /// # use core_simd::*;
+    /// let vec: Vec<i32> = vec![10, 11, 12, 13, 14, 15, 16, 17, 18];
+    /// let indices = SimdUsize::<4>::from_array([9, 3, 0, 5]);
+    ///
+    /// let result = SimdI32::<4>::gather(&vec, indices);
+    /// assert!(result == SimdI32::from_array([0, 13, 10, 15]));
+    /// ```
+    #[must_use]
+    #[inline]
+    fn gather(slice: &[Self::Scalar], idxs: SimdUsize<LANES>) -> Self
+    where
+        Self::Scalar: Default,
+    {
+        Self::gather_or(slice, idxs, Self::splat(Self::Scalar::default()))
+    }
+
+    /// SIMD gather: construct a SIMD vector by reading from a slice, using potentially discontiguous indices.
+    /// If an index is out of bounds, it instead selects from the values in the "or" vector.
+    /// ```
+    /// # use core_simd::*;
+    /// let vec: Vec<i32> = vec![10, 11, 12, 13, 14, 15, 16, 17, 18];
+    /// let indices = SimdUsize::<4>::from_array([9, 3, 0, 5]);
+    /// let alt = SimdI32::from_array([-5, -4, -3, -2]);
+    ///
+    /// let result = SimdI32::<4>::gather_or(&vec, indices, alt);
+    /// assert!(result == SimdI32::from_array([-5, 13, 10, 15]), "oh no, got {:?}", result);
+    /// ```
+    #[must_use]
+    #[inline]
+    fn gather_or(slice: &[Self::Scalar], idxs: SimdUsize<LANES>, or: Self) -> Self {
+        // IMPORTANT: We MUST construct our gather mask before we derive a pointer!
+        let mask = idxs.lanes_lt(SimdUsize::splat(slice.len())).to_int();
+        // IMPORTANT: We MUST obtain a pointer via as_ptr or an equivalent method!
+        let base_ptr = SimdConst::splat(slice.as_ptr());
+        // Ferris forgive me, I have done pointer arithmetic here.
+        let ptrs = base_ptr.wrapping_add(idxs);
+        // SAFETY: We bounds-mask the pointers so nothing fishy can emerge from this insha'allah
+        unsafe { intrinsics::simd_gather(or, ptrs, mask) }
+    }
+}
+
+macro_rules! impl_simdarray_for {
+    ($simd:ident {type Scalar = $scalar:ident;}) => {
+        impl<const LANES: usize> SimdArray<LANES> for $simd<LANES>
+            where SimdUsize<LANES>: crate::LanesAtMost32,
+            SimdIsize<LANES>: crate::LanesAtMost32,
+            MaskSize<LANES>: crate::Mask,
+            Self: crate::LanesAtMost32,
+        {
+            type Scalar = $scalar;
+
+            #[must_use]
+            #[inline]
+            fn splat(val: Self::Scalar) -> Self {
+                [val; LANES].into()
```

Comment:
This is neat!
Would I also be able to do
```rust
[1,2,3,4].into(i32x4)
```
?

Reply:
you don't pass a type into into as an argument. If you need to pick the target type, i think you can call .into::<T>() if you have to, but usually it's either inferred or you'd call T::from(val) instead.

Reply:
Oh derp, that's right.
It'd be nice to have `[1,2,3,4].into::i32x4::()` I think.

Reply:
No, `into` doesn't take an argument like that. It's relying on type inference here because Rust knows the return type is Self, so it has to be using the `impl Into<${Self}> for [T; N]`.

Reply:
And unfortunately you can't turbofish it because the generic is actually on `Into<A> for Z`, not the function, whereas `collect` can be turbofished because collect is written `fn collect<B>(self) -> B`.

Reply:
<x as Into<f32x4>>.into()

...maybe? i forget if that works